menu.py

```diff
@@ -1,38 +1,452 @@
+            break
+
+    return selected_menu, menu_path
+
+
+def mouse_pos_menu_and_path():
+    """Returns element in menu bar under cursor, and path to it"""
+    menu_path = []
+    element_at_mouse_pos = ui.element_at(*ctrl.mouse_pos())
+    if element_at_mouse_pos.AXRole in ("AXMenu", "AXMenuItem", "AXMenuBarItem"):
+        menu = element_at_mouse_pos
+        while True:
+            if menu.AXRole in ("AXMenuItem", "AXMenuBarItem"):
+                menu_path.append(menu.AXTitle)
+            menu = menu.AXParent
+            if menu.AXRole == "AXMenu":
+                menu = menu.AXParent
+            elif menu.AXRole == "AXMenuBar":
+                menu_path.reverse()
+                break
+        else:  # not in menu bar
+            return None, []
+
+    return element_at_mouse_pos, menu_path
+
+
+def selected_menu_path_strategy():
+    """Returns 'selected' element in menu bar, path to it and strategy used to find it"""
+    selected_menu, selected_menu_path = selected_menu_and_path()
+
+    if (
+        selected_menu_path
+        and hasattr(ui, "element_at")
+        and selected_menu.AXRole != "AXMenuItem"
+    ):
+        element_at_mouse_pos, mouse_pos_menu_path = mouse_pos_menu_and_path()
+        if element_at_mouse_pos and len(mouse_pos_menu_path) > len(selected_menu_path):
+            return (
+                element_at_mouse_pos,
+                mouse_pos_menu_path,
+                "Found under mouse pointer",
+            )
+
+    return selected_menu, selected_menu_path, "Found selected"
+
+
+def selected_menu_key_path_strategy():
+    """Returns Talon-format key equivalent of the menu item that is currently highlighted,
+    , path to it and strategy used to find it"""
+    selected_menu, menu_path, strategy = selected_menu_path_strategy()
+
+    if (not selected_menu) or selected_menu.AXRole != "AXMenuItem":
+        app.notify("No menu bar item selected or under the mouse pointer")
+        return None, None, None
+
+    key_char = selected_menu.get("AXMenuItemCmdChar")
+    modifiers = selected_menu.get("AXMenuItemCmdModifiers")
+    glyph = selected_menu.get("AXMenuItemCmdGlyph")
+    virtual_key = selected_menu.get("AXMenuItemCmdVirtualKey")
+
+    menu_keys = []
+
+    if modifiers is not None:
+        if not (modifiers & kMenuNoCommandModifier):
+            menu_keys.append("cmd")
+        if modifiers & kMenuShiftModifier:
+            menu_keys.append("shift")
+        if modifiers & kMenuOptionModifier:
+            menu_keys.append("alt")
+        if modifiers & kMenuControlModifier:
+            menu_keys.append("ctrl")
+        if modifiers & kMenuFnGlobeModifier:
+            menu_keys.append("fn")
+
+    got_key = False
+
+    if key_char is not None:
+        menu_keys.append(key_char.lower())
+        got_key = True
+
+    if not got_key and virtual_key is not None:
+        if key_name := VK_NAMES.get(virtual_key):
+            menu_keys.append(key_name)
+            got_key = True
```

Comment:
It feels like this chunk could be extracted to a helper function.

Reply:
Good idea — done.